main.go

```diff
@@ -80,212 +80,243 @@ func main() {
+		if err := client.CreateBackendService(bsName, negName, hcName, tagConfig.BackendService.Affinity, tagConfig.BackendService.CDN); err != nil {
+			return fmt.Errorf("Can't create backend service %q: %+v", bsName, err)
+		}
+		glog.Infof("Created backend service %q", bsName)
 
-				lock.Unlock()
-			default:
-				continue
-			}
-		case <-done:
-			glog.Warningf("Received termination signal for service with tag %s", tag)
-			wg.Done()
-			return
+		glog.Infof("Updating URL map %q...", urlMapName)
+		if err := client.UpdateURLMap(urlMapName, bsName, tagConfig.BackendService.Host, tagConfig.BackendService.Path); err != nil {
+			return fmt.Errorf("Can't update URL map %q: %+v", urlMapName, err)
+		}
+		glog.Infof("Updated URL map %q", urlMapName)
+	}
+
+	*isRunning = true
+	glog.Infof("Watching service with tag %q...", tagName)
+
+	return nil
+}
+
+func handleServiceDeletedUpdate(
+	client cloud.Cloud,
+	tagConfig *tagConfiguration,
+	lock *sync.RWMutex,
+	isRunning *bool,
+	instances map[string]*registry.ServiceInstance,
+) error {
+	lock.Lock()
+	defer lock.Unlock()
+
+	tagName := tagConfig.Name
+
+	if !*isRunning {
+		glog.Infof("Ignored DELETED update of service with tag %q since it's not running", tagName)
+		return nil
+	}
+
+	var toRemove []gce.NetworkEndpoint
+
+	for k, instance := range instances {
+		toRemove = append(toRemove, gce.NetworkEndpoint{
+			Instance: normalizeInstanceName(instance.Host),
+			IP:       instance.Address,
+			Port:     instance.Port,
+		})
+		delete(instances, k)
+	}
+
+	//do we have instances to remove from the NEG?
+	if len(toRemove) > 0 {
+		negName := tagConfig.NetworkEndpointGroup.Name
+		if err := client.DetachEndpointsFromGroup(toRemove, negName); err != nil {
+			return fmt.Errorf("Failed detaching endpoints from %q network endpoint group: %+v", negName, err)
 		}
 	}
+
+	*isRunning = false
+	return nil
 }
 
-// NOTE: Maximum length of resource name is 63 and name must start and end with letter or digit
-func makeName(prefix string, name string) string {
-	n := strings.Join([]string{prefix, name}, "-")
-	if len(n) > maximumNameLength {
-		n = strings.TrimLeft(n[:maximumNameLength], "-")
+func handleServiceChangedUpdate(
+	client cloud.Cloud,
+	tagConfig *tagConfiguration,
+	lock *sync.RWMutex,
+	isRunning *bool,
+	instances map[string]*registry.ServiceInstance,
+	update *registry.ServiceUpdate,
+) error {
+	lock.Lock()
+	defer lock.Unlock()
+
+	tagName := tagConfig.Name
+
+	if !*isRunning {
+		glog.Infof("Ignored DELETED update of service with tag %q since it's not running", tagName)
+		return nil
+	}
+
+	var toAdd, toRemove []gce.NetworkEndpoint
+
+	// finding instances to remove
+	for k, instance := range instances {
+		// Doesn't instance exist in update?
+		if _, ok := update.ServiceInstances[k]; !ok {
+			toRemove = append(toRemove, gce.NetworkEndpoint{
+				Instance: normalizeInstanceName(instance.Host),
+				IP:       instance.Address,
+				Port:     instance.Port,
+			})
+			delete(instances, k)
```

Comment:
Please don't modify the range over which you are iterating in the iterating cycle itself.